### PR TITLE
test: add and cleanup vm-console-proxy operand

### PR DESCRIPTION
**Pull request by codingben**
**What this PR does / why we need it**:

Add and cleanup operand unit tests.

**Release note**:

```release-note
Add and cleanup vm-console-proxy operand unit tests
```

### CODE REVIEW

**Comment by codingben** [comment]
@akrejcir FYI, it can be reviewed except an issue in `should create vm-console-proxy resources`.

File: internal/operands/vm-console-proxy/resources.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
+package vm_console_proxy
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/resources.go:1]: The resources for `vm-console-proxy` are loaded from a yaml file, so they should not be hardcoded here.
I think you can remove this file.

File: internal/operands/vm-console-proxy/reconcile.go [hunk 2]
```diff
@@ -203,7 +203,9 @@ func reconcileDeploymentFuncs(deployment apps.Deployment) common.ReconcileFunc {
 			ClusterResource(&deployment).
 			WithAppLabels(operandName, operandComponent).
 			UpdateFunc(func(newRes, foundRes client.Object) {
-				foundRes.(*apps.Deployment).Spec = newRes.(*apps.Deployment).Spec
+				newDeployment := newRes.(*apps.Deployment)
+				foundDeployment := foundRes.(*apps.Deployment)
+				foundDeployment.Spec = newDeployment.Spec
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile.go:208]: This change is not needed. The one line was clear enough.

File: internal/operands/vm-console-proxy/reconcile.go [hunk 3]
```diff
@@ -228,30 +230,28 @@ func reconcileDeploymentFuncs(deployment apps.Deployment) common.ReconcileFunc {
...
-	if err != nil {
-		return false
+	if request.Instance.GetAnnotations() != nil {
+		enable, isFound := request.Instance.GetAnnotations()[EnableAnnotation]
+		if isFound {
+			isEnabled, err := strconv.ParseBool(enable)
+			if err == nil {
+				return isEnabled
+			}
+		}
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile.go:240]: I think this change is not needed. Early `return`s are easier to read than nested `if`s.

File: internal/operands/vm-console-proxy/reconcile.go [hunk 4]
```diff
@@ -228,30 +230,28 @@ func reconcileDeploymentFuncs(deployment apps.Deployment) common.ReconcileFunc {
...
-		return defaultNamespace
-	}
-	namespaceStr, ok := request.Instance.GetAnnotations()[VmConsoleProxyNamespaceAnnotation]
-	if !ok {
-		return defaultNamespace
+	if request.Instance.GetAnnotations() != nil {
+		namespace, isFound := request.Instance.GetAnnotations()[VmConsoleProxyNamespaceAnnotation]
+		if isFound {
+			return namespace
+		}
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile.go:251]: This change is ok, I think.

**Review by akrejcir — COMMENTED** [review]
Can you also cleanup the names of functions?

File: internal/operands/vm-console-proxy/reconcile.go (hunk 3 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile.go:240]: I agree, but isn't makes us to add another unit test, for that early return?

Comment (akrejcir) [reply internal/operands/vm-console-proxy/reconcile.go:262]: In general, unit tests should check the **functionality** of a code "unit", usually a `struct` and it's public methods.
So a code change that does not change functionality, should not change unit tests. In fact unit tests are useful to verify that the changed code has the same functionality.

File: internal/operands/vm-console-proxy/reconcile.go [hunk 5]
```diff
@@ -252,30 +252,28 @@ func reconcileRoute(serviceName string) common.ReconcileFunc {
...
 func getVmConsoleProxyNamespace(request *common.Request) string {
-	const defaultNamespace = "kubevirt"
-	if request.Instance.GetAnnotations() == nil {
-		return defaultNamespace
-	}
-	namespaceStr, ok := request.Instance.GetAnnotations()[VmConsoleProxyNamespaceAnnotation]
-	if !ok {
-		return defaultNamespace
+	if request.Instance.GetAnnotations() != nil {
+		namespace, isFound := request.Instance.GetAnnotations()[VmConsoleProxyNamespaceAnnotation]
```
Comment (0xFelix) [inline internal/operands/vm-console-proxy/reconcile.go:270]: You could merge L270 and 271.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 6]
```diff
@@ -54,31 +56,236 @@ var _ = Describe("VM Console Proxy Operand", func() {
...
 	})
 
 	It("should return functions from reconcile correclty", func() {
-		functions, err := v.Reconcile(mockedRequest)
+		functions, err := operand.Reconcile(&request)
 		Expect(err).ToNot(HaveOccurred(), "should not throw err")
 		Expect(len(functions)).To(Equal(7), "should return correct number of reconcile functions")
 	})
+
+	It("should create vm-console-proxy resources", func() {
```
Comment (0xFelix) [inline internal/operands/vm-console-proxy/reconcile_test.go:69]: Strictly speaking this is not cleaning up existing code but adding new test code. Could you please put it into a separate commit?

Comment (akrejcir) [reply internal/operands/vm-console-proxy/reconcile_test.go:69]: Ok, then please use this PR to add new unit tests and you can drop the refactoring changes.

Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:69]: Currently some new unit tests depend one refactored functions, `isEnabled` and `getVmConsoleProxyNamespace`. Otherwise, unit tests won't fit there.

File: internal/operands/vm-console-proxy/reconcile.go (hunk 5 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile.go:270]: Do you mean:
```
if namespace, isFound := request.Instance.GetAnnotations()[VmConsoleProxyNamespaceAnnotation]; isFound {
			return namespace
		}
```
If yes, is it common code style? Because I don't like it and feel it's line too long for good code readability.

Comment (akrejcir) [reply internal/operands/vm-console-proxy/reconcile.go:270]: It is common, but It's up to you to decide. We don't have guidelines about this.

**Comment by codingben** [comment]
I'll rename that PR to add only new tests, and open another PR to have refactored functions.

File: internal/operands/vm-console-proxy/resources.go (hunk 1 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/resources.go]: Done, thanks!

**Comment by codingben** [comment]
This PR is ready for review, except some issue with the commented lines:
```
It("should remove cluster resources on cleanup", func() {
...
		// ExpectResourceNotExists(bundle.serviceAccount, request)
		ExpectResourceNotExists(bundle.clusterRole, request)
		ExpectResourceNotExists(bundle.clusterRoleBinding, request)
		// ExpectResourceNotExists(bundle.configMap, request)
		// ExpectResourceNotExists(bundle.service, request)
		// ExpectResourceNotExists(bundle.deployment, request)
		ExpectResourceNotExists(newRoute(namespace, "vm-console-proxy"), request)
}
```
It throws nil error when I try to test it.

/cc @akrejcir @0xFelix

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 7]
```diff
@@ -17,13 +18,17 @@ import (
...
+	. "kubevirt.io/ssp-operator/internal/test-utils"
 	vm_console_proxy_bundle "kubevirt.io/ssp-operator/internal/vm-console-proxy-bundle"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 const (
+	containerPort          = 8768
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:30]: This const is not used.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 8]
```diff
@@ -35,12 +40,20 @@ const (
...
 
 var _ = Describe("VM Console Proxy Operand", func() {
-	var v *vmConsoleProxy
-	var mockedRequest *common.Request
+	const (
+		replicas int32 = 1
+	)
+
+	var (
+		bundle  *vmConsoleProxy
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:48]: `bundle` should have type `*vm_console_proxy_bundle.Bundle`

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 9]
```diff
@@ -35,12 +40,20 @@ const (
...
+		operand *vmConsoleProxy
+		request common.Request
+	)
 
 	BeforeEach(func() {
-		v = getMockedVmConsoleProxyOperand()
-		mockedRequest = getMockedRequest()
+		bundle = New(getMockedTestBundle())
+		operand = getMockedVmConsoleProxyOperand()
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:55]: Please use `New()` to create the operand. It doesn't make sense to mock and object that is being tested.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 10]
```diff
@@ -35,12 +40,20 @@ const (
...
+		request = getMockedRequest()
 	})
 
 	It("should return new with test bundle correctly", func() {
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:59]: Please remove this test, because it only tests private fields. Usually unit tests are only for exported functionality.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 11]
```diff
@@ -54,30 +67,234 @@ var _ = Describe("VM Console Proxy Operand", func() {
...
+
+		// All resources should be available
+		for _, reconcileResult := range reconcileResults {
+			Expect(reconcileResult.Status.NotAvailable).To(BeNil())
+			Expect(reconcileResult.Status.Progressing).To(BeNil())
+			Expect(reconcileResult.Status.Degraded).To(BeNil())
+		}
+	})
+
+	Describe("GetVmConsoleProxyImage", func() {
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:188]: Unit tests should not test private functions. Can you rewrite this in terms of the created deployment?
Or you can just remove it.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 12]
```diff
@@ -54,30 +67,234 @@ var _ = Describe("VM Console Proxy Operand", func() {
...
+		})
+
+		It("should return the default value if VM_CONSOLE_PROXY_IMAGE not set", func() {
+			defaultImage := "quay.io/kubevirt/vm-console-proxy:v0.1.0"
+			actualImage := getVmConsoleProxyImage()
+			Expect(actualImage).To(Equal(defaultImage))
+		})
+	})
+
+	Describe("GetVmConsoleProxyNamespace", func() {
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:204]: Similarly, these tests also call private function.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 13]
```diff
@@ -54,30 +67,234 @@ var _ = Describe("VM Console Proxy Operand", func() {
...
+						},
+					},
+				},
+			}
+			actualNamespace := getVmConsoleProxyNamespace(request)
+			Expect(actualNamespace).To(Equal(customNamespace))
+		})
+	})
+
+	Describe("IsEnabled", func() {
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:245]: The same here.

**Review by akrejcir — COMMENTED** [review]
A partial review. 
In general, unit tests don't need to test every single private function. Private functions can change. Tests should cover visible behavior of a module:
- Exported functions and methods
- Exported structs
- Exported fields of structs

In my opinion, an exception to this guideline is if a private function has complex implementation and we want to make sure that it works.

**Comment by codingben** [comment]
@akrejcir Ah okay :) I'm so sorry. I never wrote before operator tests. I thought we should test all operand functions because it plays critical role in deploying `vm-console-proxy`, it's not just another library. Thanks!

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 14]
```diff
@@ -54,30 +67,234 @@ var _ = Describe("VM Console Proxy Operand", func() {
...
+		ExpectResourceExists(bundle.clusterRoleBinding, request)
+		ExpectResourceExists(bundle.configMap, request)
+		ExpectResourceExists(bundle.service, request)
+		ExpectResourceExists(bundle.deployment, request)
+		ExpectResourceExists(newRoute(namespace, "vm-console-proxy"), request)
+
+		_, err = operand.Cleanup(&request)
+		Expect(err).ToNot(HaveOccurred())
+
+		// ExpectResourceNotExists(bundle.serviceAccount, request)
```
Comment (0xFelix) [inline internal/operands/vm-console-proxy/reconcile_test.go:108]: Can you please investigate what makes these calls fail? Commented code IMO is unclean.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 15]
```diff
@@ -54,30 +67,234 @@ var _ = Describe("VM Console Proxy Operand", func() {
...
+		_, err := operand.Reconcile(&request)
+		Expect(err).ToNot(HaveOccurred())
+
+		ExpectResourceExists(bundle.serviceAccount, request)
+		ExpectResourceExists(bundle.clusterRole, request)
+		ExpectResourceExists(bundle.clusterRoleBinding, request)
+		ExpectResourceExists(bundle.configMap, request)
+		ExpectResourceExists(bundle.service, request)
+		ExpectResourceExists(bundle.deployment, request)
+		ExpectResourceExists(newRoute(namespace, "vm-console-proxy"), request)
```
Comment (0xFelix) [inline internal/operands/vm-console-proxy/reconcile_test.go:103]: You might use a const for the route name in this spec.

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 6 above)
Comment (0xFelix) [reply internal/operands/vm-console-proxy/reconcile_test.go:80]: IMO if a PR gets large, multiple commits help to reduce the cognitive load of reviewing the PR significantly. Too many changes in a single commit make PRs very hard to review.

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 14 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:108]: Yep, thanks to Andrej's review, I saw that I was using wrong resource reference.

**Comment by codingben** [comment]
@akrejcir @0xFelix Thanks a lot for your reviews, really appreciated. It should be okay now. :)

**Comment by codingben** [comment]
> Can you also cleanup the names of functions?

Sorry, what do you mean?

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 16]
```diff
@@ -35,49 +38,153 @@ const (
...
+	It("should create vm-console-proxy resources", func() {
+		_, err := operand.Reconcile(&request)
+		Expect(err).ToNot(HaveOccurred())
+
+		ExpectResourceExists(operand.serviceAccount, request)
+		ExpectResourceExists(operand.clusterRole, request)
+		ExpectResourceExists(operand.clusterRoleBinding, request)
+		ExpectResourceExists(operand.configMap, request)
+		ExpectResourceExists(operand.service, request)
+		ExpectResourceExists(operand.deployment, request)
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:75]: It's better if these tests check the resources from the mocked bundle, and don't access private fields of the operand. Because private fields can change, and unit tests should not depend on them.

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 15 above)
Comment (akrejcir) [reply internal/operands/vm-console-proxy/reconcile_test.go:89]: +1

**Comment by akrejcir** [comment]
> > Can you also cleanup the names of functions?
> 
> @akrejcir Sorry, what do you mean?

I meant that function names in `reconcile.go` can be made clearer. For example `reconcileServiceAccountsFuncs()` can just be `reconcileServiceAccounts()`. But now that you've split this PR into 2, that comment is relevant to the other PR.

**Comment by akrejcir** [comment]
Can you add unit tests for the annotations functionality? To test these cases:
- proxy is deployed to the namespace specified by the `vm-console-proxy-namespace` annotation
- when `vm-console-proxy-enabled` annotation is removed, the proxy resources are deleted

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 15 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:89]: @akrejcir I already changed it. The parameter `vm-console-proxy` in `newRoute(namespace string, serviceName string)` isn't route name but service name.

Comment (akrejcir) [reply internal/operands/vm-console-proxy/reconcile_test.go]: Right, sorry. I just realized that service name is irrelevant for this test.

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 16 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:75]: What do you mean by "resources from the mocked bundle", by loading it from `vm-console-proxy-bundle-test.yaml` or just a new function?

Comment (akrejcir) [reply internal/operands/vm-console-proxy/reconcile_test.go:75]: By "mocked bundle", I mean the `*vm_console_proxy_bundle.Bundle` returned by `getMockedTestBundle()`.
So for example, in `BeforeEach()` you could store the bundle into a variable, and then here check that the resource from the bundle was created by the operand:
```go
ExpectResourceExists(bundle.Deployment, request)
```

This way the test does not access any private fields.

Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:75]: @akrejcir Yes, got it. I think this is what I did initially but then changed it yesterday.

Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go]: Done.

**Comment by codingben** [comment]
> proxy is deployed to the namespace specified by the vm-console-proxy-namespace annotation

Done, please review.

> when vm-console-proxy-enabled annotation is removed, the proxy resources are deleted

@akrejcir Done, please review as well.

But, what is the purpose of that test? Isn't similar to `should remove cluster resources on cleanup`? I mean, it sounds like it's the same test as `should remove cluster resources on cleanup` and we just need to remove annotation by:
```
request.Instance.Annotations = map[string]string{
    VmConsoleProxyNamespaceAnnotation: namespace,
}
```
Before we call to Cleanup?

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 17]
```diff
@@ -35,49 +38,187 @@ const (
...
+			Expect(reconcileResult.Status.Progressing).To(BeNil())
+			Expect(reconcileResult.Status.Degraded).To(BeNil())
+		}
+	})
+
+	It("should deploy resources in namespace provided by annotation", func() {
+		_, err := operand.Reconcile(&request)
+		Expect(err).ToNot(HaveOccurred())
+
+		namespace := request.Instance.GetAnnotations()[VmConsoleProxyNamespaceAnnotation]
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:180]: Instead of getting the namespace, can you change the value of the annotation to something different?
And then after calling `operand.Reconcile()`, test that the resources are deployed in that namespace.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 18]
```diff
@@ -35,49 +38,187 @@ const (
...
+		ExpectNamespacedResourceExists(&bundle.ServiceAccount, request, namespace)
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:181]: Function `ExpectNamespacedResourceExists()` is not defined anywhere.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 19]
```diff
@@ -35,49 +38,187 @@ const (
...
+		ExpectNamespacedResourceExists(newRoute(getVmConsoleProxyNamespace(&request), serviceName), request, namespace)
+	})
+
+	It("should delete resources when enabled annotation is removed", func() {
+		_, err := operand.Reconcile(&request)
+		Expect(err).ToNot(HaveOccurred())
+
+		request.Instance.Annotations = map[string]string{
+			VmConsoleProxyNamespaceAnnotation: namespace,
+		}
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:194]: Instead of setting the annotation map, can you just delete the `EnableAnnotation`?
```go
delete(request.Instance.Annotations, EnableAnnotation)
```

Then it will be more obvious, that this test removes an annotation.

File: internal/operands/vm-console-proxy/reconcile_test.go [hunk 20]
```diff
@@ -35,49 +38,187 @@ const (
...
+
+		_, err = operand.Cleanup(&request)
+		Expect(err).ToNot(HaveOccurred())
```
Comment (akrejcir) [inline internal/operands/vm-console-proxy/reconcile_test.go:197]: `Cleanup()` is normally called when the SSP CR is removed from the cluster.
This test only tests removing an annotation, not the whole CR, so it should call `Reconcile()` here.

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 18 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:181]: It wasn't added to the commit, but `ExpectNamespacedResourceExists is part of test_utils.go`.

File: internal/test-utils/test_utils.go [hunk 21]
```diff
@@ -19,3 +19,10 @@ func ExpectResourceNotExists(resource client.Object, request common.Request) {
 	ExpectWithOffset(1, err).To(HaveOccurred())
 	ExpectWithOffset(1, errors.IsNotFound(err)).To(BeTrue())
 }
+
+func ExpectNamespacedResourceExists(resource client.Object, request common.Request, namespace string) {
+	key := client.ObjectKeyFromObject(resource)
+	err := request.Client.Get(request.Context, key, resource)
+	ExpectWithOffset(1, err).ToNot(HaveOccurred())
```
Comment (0xFelix) [inline internal/test-utils/test_utils.go:26]: IMO either use `...WithOffset` in all assertions or in none. I'd vote for using the standard `Expect`.

Comment (akrejcir) [reply internal/test-utils/test_utils.go:26]: +1 for the consistency. 
But in this case using no offset will generate useless error message, because it will point to this line without any information about the calling function or the values of parameters passed.

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 17 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:180]: Since I'm doing: `ExpectNamespacedResourceExists(&bundle.Deployment, request, namespace)` it doesn't work, I think this is because `&bundle.Deployment` contains value of `namespace` from constant variable.

Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:180]: Then we should do: `ExpectNamespacedResourceExists(operand.serviceAccount, request, namespace)` to use changed resource in `operand.serviceAccount`?

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 19 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:194]: Done, thanks!

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 20 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go]: Done, thanks! :)

File: internal/operands/vm-console-proxy/reconcile_test.go (hunk 17 above)
Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go:180]: I'm trying to do it, but for some reason it doesn't work, I'm spending some time to debug it.

Comment (codingben) [reply internal/operands/vm-console-proxy/reconcile_test.go]: Done, thanks @akrejcir.

**Comment by codingben** [comment]
@akrejcir Seems like approval is missing and it wasn't merged because of it.
